### ISSUE TITLE
Allow configuration of parser

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#1624]
Many Esprima-based tools allow the end-user to specify an alternate parser to use (so long as they are Esprima compatible). We should do the same for  ESLint. This would allow, for example, someone to specify to use Esprima-FB instead of Espree.

I'm unsure the best way to do this currently. My first thought was to have a `--parser` flag on the command line, but that would require someone getting that info into the `eslint` object, which isn't supposed to know about the command line. The other thought was to put it in `settings`, such as:

```json
{
  "settings": {
    "parser": "esprima-fb"
  }
}
```

At the moment, I'm leaning towards the `settings` approach, since it would also allow rules to determine if the regular parser was being used or not.

**Comment by btmills** [comment]
Since both this and #1602 deal with the parser and its options, it seems it would make sense to come up with something that might keep these together. Just some ideas:

```js
{
  "settings": {

    // Custom parser with default options
    "parser": "esprima-fb",

    // Default parser with custom options
    "parser": {
      "ecmaFeatures": {
        "generators": true,
        "jsx": true
      }
    },

    // Custom parser with custom options
    "parser": {
      "module": "acorn",
      "options": {
        "ecmaVersion": 6
      }
    }

  }
}
```

**Comment by nzakas** [comment]
I'm not sure it makes sense to combine these, as I'd prefer people think about configuring ESLint with certain features rather than configuring the underlying parser directly, nor am I sure I want to allow people to configure an external parser. This is a dicey feature as it is, so I'd like to keep it as simple as possible.

**Comment by yannickcr** [comment]
I opened a related issue in JSCS https://github.com/jscs-dev/node-jscs/issues/865 that allows us to use a custom parser but not to configure it which is very limiting.

**Comment by nzakas** [comment]
Configuring is more complex and we don't have any use cases for that right now. We do have a request to allow the usage of esprima-fb directly, and that's what I'd like to support.

**Comment by nzakas** [comment]
Working on this.
